DistroLauncher/WinOobeStrategy.cpp

```diff
@@ -0,0 +1,318 @@
+                      "onGui callback must be a callable returning HRESULT with no arguments");
+        static_assert(std::is_convertible_v<OnTui, std::function<HRESULT()>>,
+                      "onTui callback must be a callable returning HRESULT with no arguments");
+        static_assert(std::is_convertible_v<OnNone, std::function<HRESULT()>>,
+                      "onNone callback must be a callable returning HRESULT with no arguments");
+
+        // No further checks if LAUNCHER_FORCE_MODE is set.
+        LauncherForceMode forced = environmentForceMode();
+        switch (forced) {
+        case LauncherForceMode::Invalid:
+            [[fallthrough]];
+        case LauncherForceMode::Unset:
+            break;
+        case LauncherForceMode::TextForced:
+            return onTui();
+        case LauncherForceMode::GuiForced:
+            return onGui();
+        }
+
+        // WSL 1 cannot mount the Subiquity snap, thus neither GUI nor TUI will be available.
+        if (Oobe::internal::WslGetDistroSubsystemVersion() < 2) {
+            return onNone();
+        }
+        // if none of the required snaps exist in the rootfs no UI will be available.
+        if (!internal::hasUdiSnap() && !internal::hasSubiquitySnap()) {
+            return onNone();
+        }
+
+        // if there is no way to boostrap the OOBE snap, no UI available.
+        if (!DistributionInfo::isOOBEAvailable()) {
+            return onNone();
+        }
+
+        // Without localhost forwarding, the OOBE will not be able to connect to the server.
+        // TUI mode is still supported.
+        if (!Oobe::internal::isLocalhostForwardingEnabled(getWslConfigPath())) {
+            return onTui();
+        }
+
+        // Otherwise, running the OOBE on Windows is supported. Platform is implied, since this strategy is selected at
+        // compile time based on whether the platform supports running the GUI app.
+        return onGui();
+    }
+
+    bool WslFileBuf::isEmpty() const
+    {
+        return contents.empty();
+    }
+
+    bool WslFileBuf::write() const
+    {
+        const auto path = WindowsPath(linuxPath);
+        std::ofstream file;
+        file.open(path.string(), std::ios::ate);
+        if (file.fail()) {
+            return false;
+        }
+        file << contents;
+        file.close();
+        return !file.fail();
+    }
+
+    HRESULT Oobe::WinOobeStrategy::do_tui_reconfigure()
+    {
+        return Oobe::internal::reconfigure_linux_ui(installer);
+    }
+
+    HRESULT WinOobeStrategy::do_gui_reconfigure()
+    {
+        // The OOBE process must not be previously created.
+        assert(!oobeProcess.has_value());
+        const auto cli = makeCli(L"--reconfigure");
+        oobeProcess.emplace(oobeExePath, cli.c_str(), nullptr, nullptr, nullptr);
+        if (!oobeProcess.has_value()) {
+            return E_UNEXPECTED;
+        }
+        auto& oobe = oobeProcess.value();
+        if (!oobe.start()) {
```

Comment:
You can use `obeeProcess->start()` if you're doing this to avoid calling `.value()` every time.

Reply:
Good point. I think it's clearer, even though it looks like a pointer.